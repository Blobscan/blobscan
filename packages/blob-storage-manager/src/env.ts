import {
  z,
  booleanSchema,
  createEnv,
  presetEnvOptions,
  maskSensitiveData,
} from "@blobscan/zod";

export function parseEnv() {
  return createEnv({
    envOptions: {
      server: {
        BEE_DEBUG_ENDPOINT: z.string().url().optional(),
        BEE_ENDPOINT: z.string().url().optional(),
        CHAIN_ID: z.coerce.number().positive().default(1),
        FILE_SYSTEM_STORAGE_ENABLED: booleanSchema.default("false"),
        FILE_SYSTEM_STORAGE_PATH: z.string().optional(),
        GOOGLE_STORAGE_BUCKET_NAME: z.string().optional(),
        GOOGLE_STORAGE_PROJECT_ID: z.string().optional(),
        GOOGLE_SERVICE_KEY: z.string().optional(),
        GOOGLE_STORAGE_API_ENDPOINT: z.string().optional(),
        GOOGLE_STORAGE_ENABLED: booleanSchema.default("false"),
        POSTGRES_STORAGE_ENABLED: booleanSchema.default("true"),
        SWARM_STORAGE_ENABLED: booleanSchema.default("false"),
      },

      ...presetEnvOptions,
    },
    display(env) {
      console.log(
        `Blob storage manager configuration: chainId=${env.CHAIN_ID}, file_system=${env.FILE_SYSTEM_STORAGE_ENABLED} postgres=${env.POSTGRES_STORAGE_ENABLED}, gcs=${env.GOOGLE_STORAGE_ENABLED}, swarm=${env.SWARM_STORAGE_ENABLED}`
      );

      if (env.GOOGLE_STORAGE_ENABLED) {
        console.log(
          `GCS configuration: bucketName=${
            env.GOOGLE_STORAGE_BUCKET_NAME
          }, projectId=${maskSensitiveData(
            env.GOOGLE_STORAGE_PROJECT_ID
          )}, apiEndpoint=${env.GOOGLE_STORAGE_API_ENDPOINT}`
        );
      }

      if (env.SWARM_STORAGE_ENABLED) {
        console.log(
          `Swarm configuration: beeEndpoint=${env.BEE_ENDPOINT}, debugEndpoint=${env.BEE_DEBUG_ENDPOINT}`
        );
      }

      if (env.FILE_SYSTEM_STORAGE_ENABLED) {
        console.log(
          `File system configuration: blobDirPath=${env.FILE_SYSTEM_STORAGE_PATH}`
        );
      }
    },
  });
}
export const env = parseEnv();

export type Environment = typeof env;
