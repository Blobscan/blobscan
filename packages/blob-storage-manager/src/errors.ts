class ErrorException extends Error {
  constructor(message: string, cause?: Error | Error[]) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class BlobStorageError extends ErrorException {
  constructor(storageName: string, message: string, cause?: Error) {
    super(`${storageName} failed: ${message}`, cause);
  }
}

export class StorageCreationError extends BlobStorageError {
  constructor(storageName: string, message: string, cause: Error) {
    super(
      storageName,
      `Creation failed${message.length ? `: ${message}` : ""}`,
      cause
    );
  }
}

export class BlobStorageManagerError extends ErrorException {
  constructor(message: string, cause?: Error | BlobStorageError[]) {
    super(message, cause);
  }
}
