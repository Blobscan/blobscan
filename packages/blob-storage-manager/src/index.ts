export { BlobStorageManager } from "./BlobStorageManager";
export * from "./blob-storage-manager";
export { env } from "./env";
export type { Environment } from "./env";
export * from "./errors";
export { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";
export * from "./types";
