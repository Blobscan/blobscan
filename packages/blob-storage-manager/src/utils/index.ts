export * from "./blob";
export * from "./file";
export * from "./storage";
