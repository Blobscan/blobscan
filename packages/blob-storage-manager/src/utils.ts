import { $Enums } from "@blobscan/db";

import type { BlobStorage } from "./BlobStorage";
import { env } from "./env";
import type { StorageCreationError } from "./errors";
import {
  FileSystemStorage,
  GoogleStorage,
  PostgresStorage,
  SwarmStorage,
} from "./storages";
import type { BlobStorageName } from "./types";

export const BLOB_STORAGE_NAMES = $Enums.BlobStorage;

export function removeDuplicatedStorages(
  blobStorages: BlobStorage[]
): BlobStorage[] {
  return blobStorages.filter(
    (blobStorage, index, self) =>
      index === self.findIndex((t) => t.name === blobStorage.name)
  );
}

export function calculateBlobBytes(blob: string): number {
  return blob.slice(2).length / 2;
}

export async function createStorageFromEnv(
  storageName: BlobStorageName
): Promise<[BlobStorage | undefined, StorageCreationError | undefined]> {
  let storage: BlobStorage | undefined;
  let storageError: StorageCreationError | undefined;

  switch (storageName) {
    case BLOB_STORAGE_NAMES.GOOGLE:
      [storage, storageError] = await GoogleStorage.tryCreateFromEnv(env);
      break;
    case BLOB_STORAGE_NAMES.POSTGRES:
      [storage, storageError] = await PostgresStorage.tryCreateFromEnv(env);
      break;
    case BLOB_STORAGE_NAMES.SWARM:
      [storage, storageError] = await SwarmStorage.tryCreateFromEnv(env);
      break;
    case BLOB_STORAGE_NAMES.FILE_SYSTEM:
      [storage, storageError] = await FileSystemStorage.tryCreateFromEnv(env);
      break;
  }

  return [storage, storageError];
}
