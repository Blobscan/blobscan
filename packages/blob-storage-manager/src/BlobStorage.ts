import type { Environment } from "./env";
import { BlobStorageError, StorageCreationError } from "./errors";
import type { BlobStorageName } from "./types";

export interface BlobStorageConfig {
  chainId: number;
}

export abstract class BlobStorage {
  chainId: number;
  name: BlobStorageName;

  constructor(name: BlobStorageName, chainId: number) {
    this.name = name;
    this.chainId = chainId;
  }

  protected abstract _healthCheck(): Promise<void>;
  protected abstract _getBlob(uri: string): Promise<string>;
  protected abstract _storeBlob(hash: string, data: string): Promise<string>;

  async healthCheck(): Promise<"OK"> {
    try {
      await this._healthCheck();

      return "OK";
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        "Storage is not reachable",
        err as Error
      );
    }
  }

  async getBlob(uri: string): Promise<string> {
    try {
      const blob = await this._getBlob(uri);

      return blob;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to get blob with uri "${uri}"`,
        err as Error
      );
    }
  }

  async storeBlob(hash: string, data: string): Promise<string> {
    try {
      const res = await this._storeBlob(hash, data);

      return res;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to store blob with hash "${hash}"`,
        err as Error
      );
    }
  }

  protected buildBlobFileName(hash: string): string {
    return `${this.chainId.toString()}/${hash.slice(2, 4)}/${hash.slice(
      4,
      6
    )}/${hash.slice(6, 8)}/${hash.slice(2)}.txt`;
  }

  static async create<T extends BlobStorage, C extends BlobStorageConfig>(
    this: new (config: C) => T,
    config: C
  ): Promise<T> {
    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return blobStorage;
    } catch (err) {
      const err_ = err as Error;
      throw new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
    }
  }

  static async tryCreateFromEnv<
    C extends BlobStorageConfig,
    T extends BlobStorage
  >(
    this: {
      new (config: C): T;
      getConfigFromEnv(env: Partial<Environment>): C;
    },
    env: Partial<Environment>
  ): Promise<[T?, StorageCreationError?]> {
    let config: C;

    try {
      config = this.getConfigFromEnv(env);
    } catch (err) {
      const creationError = new StorageCreationError(
        this.name,
        "Failed to get config",
        err as Error
      );
      return [, creationError];
    }

    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return [blobStorage];
    } catch (err) {
      const err_ = err as Error;
      const creationError = new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
      return [, creationError];
    }
  }

  protected static getConfigFromEnv(
    env: Partial<Environment>
  ): BlobStorageConfig {
    if (!env.CHAIN_ID) {
      throw new BlobStorageError(
        this.name,
        `No config variables found: no chain id provided`
      );
    }

    return {
      chainId: env.CHAIN_ID,
    };
  }
}
