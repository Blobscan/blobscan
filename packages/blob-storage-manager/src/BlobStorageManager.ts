import { api, SemanticAttributes } from "@blobscan/open-telemetry";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManagerError } from "./errors";
import type { BlobStorageError } from "./errors";
import { tracer, updateBlobStorageMetrics } from "./instrumentation";
import type {
  Blob,
  BlobReference,
  BlobStorageName,
  StorageOf,
  StoreOptions,
} from "./types";
import { calculateBlobBytes, removeDuplicatedStorages } from "./utils";

export class BlobStorageManager {
  #blobStorages: BlobStorage[];

  constructor(blobStorages: BlobStorage[]) {
    if (!blobStorages.length) {
      throw new BlobStorageManagerError("No blob storages provided");
    }

    this.#blobStorages = removeDuplicatedStorages(blobStorages);
  }

  getStorage<T extends BlobStorageName>(name: T): StorageOf<T> | undefined {
    const blobStorage = this.#blobStorages.find(
      (storage): storage is StorageOf<T> => storage.name === name
    );

    return blobStorage as StorageOf<T> | undefined;
  }

  hasStorage<T extends BlobStorageName>(name: T): boolean {
    return !!this.getStorage(name);
  }

  async getBlob(
    ...blobReferences: BlobReference<BlobStorageName>[]
  ): Promise<{ data: string; storage: BlobStorageName } | null> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "getBlob",
        },
      },
      async (span) => {
        const availableReferences = blobReferences.filter(({ storage }) =>
          this.hasStorage(storage)
        );

        try {
          const blob = await Promise.any(
            availableReferences.map(({ reference, storage: storageName }) =>
              tracer.startActiveSpan(
                "blob_storage_manager:storage",
                {
                  attributes: {
                    storage: storageName,
                    [SemanticAttributes.CODE_FUNCTION]: "getBlob",
                  },
                },
                async (storageSpan) => {
                  const start = performance.now();
                  const storage = this.getStorage(storageName);

                  if (!storage) {
                    throw new BlobStorageManagerError(
                      `Storage "${storageName}" is not available`
                    );
                  }

                  const blob = await storage
                    .getBlob(reference)
                    .then((data) => ({
                      data,
                      storage: storageName,
                    }));
                  const end = performance.now();

                  storageSpan.end();

                  updateBlobStorageMetrics({
                    storage: storageName,
                    blobSize: calculateBlobBytes(blob.data),
                    direction: "received",
                    duration: end - start,
                  });

                  return blob;
                }
              )
            )
          );

          span.end();

          return blob;
        } catch (e) {
          const message = "Failed to get blob from any of the storages";
          let err: BlobStorageManagerError;

          // Aggregate errors are thrown when all the promises fail
          if (e instanceof AggregateError) {
            err = new BlobStorageManagerError(
              message,
              e.errors as BlobStorageError[]
            );
          } else {
            err = new BlobStorageManagerError(message, e as Error);
          }

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);

          throw err;
        } finally {
          span.end();
        }
      }
    );
  }

  async storeBlob(
    { data, versionedHash }: Blob,
    opts?: StoreOptions
  ): Promise<{
    references: BlobReference<BlobStorageName>[];
    errors: BlobStorageError[];
  }> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
        },
      },
      async (span) => {
        const selectedStorages = this.#getSelectedStorages(opts);

        const results = await Promise.allSettled(
          selectedStorages.map((storage) => {
            return tracer.startActiveSpan(
              "blob_storage_manager:storage",
              {
                attributes: {
                  storage: storage.name,
                  [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
                },
              },
              async (storageSpan) => {
                const start = performance.now();
                const blobReference = await storage
                  .storeBlob(versionedHash, data)
                  .then<BlobReference<BlobStorageName>>((reference) => ({
                    reference,
                    storage: storage.name,
                  }));
                const end = performance.now();

                storageSpan.end();

                updateBlobStorageMetrics({
                  blobSize: calculateBlobBytes(data),
                  direction: "sent",
                  duration: end - start,
                  storage: storage.name,
                });

                return blobReference;
              }
            );
          })
        );
        const references = results
          .filter(
            (
              res
            ): res is PromiseFulfilledResult<BlobReference<BlobStorageName>> =>
              res.status === "fulfilled"
          )
          .map((res) => res.value);
        const storageErrors = results
          .filter((r) => r.status === "rejected")
          .map((r) => (r as PromiseRejectedResult).reason as BlobStorageError);

        if (!references.length) {
          const err = new BlobStorageManagerError(
            `Failed to upload blob with hash "${versionedHash}" to any storage`,
            storageErrors
          );

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);
          throw err;
        }

        span.end();

        return {
          references,
          errors: storageErrors,
        };
      }
    );
  }

  #getSelectedStorages(
    { selectedStorages }: StoreOptions = { selectedStorages: [] }
  ): BlobStorage[] {
    const uniqueInputStorageNames = Array.from(new Set(selectedStorages));
    const selectedAvailableStorages = uniqueInputStorageNames.length
      ? this.#blobStorages.filter((s) =>
          uniqueInputStorageNames.includes(s.name)
        )
      : // Use all available storages if none are selected
        this.#blobStorages;

    if (
      uniqueInputStorageNames.length &&
      selectedAvailableStorages.length !== uniqueInputStorageNames.length
    ) {
      const selectdAvailableStorageNames = selectedAvailableStorages.map(
        (s) => s.name
      );
      const missingStorageNames = uniqueInputStorageNames.filter(
        (storageName) => !selectdAvailableStorageNames.includes(storageName)
      );

      throw new BlobStorageManagerError(
        `Some of the selected storages are not available: ${missingStorageNames.join(
          ", "
        )}"`
      );
    }

    return selectedAvailableStorages;
  }
}
