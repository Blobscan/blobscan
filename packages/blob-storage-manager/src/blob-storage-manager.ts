import { logger } from "@blobscan/logger";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManager } from "./BlobStorageManager";
import { env } from "./env";
import type { Environment } from "./env";
import type { BlobStorageName } from "./types";
import { BLOB_STORAGE_NAMES, createStorageFromEnv } from "./utils";

let blobStorageManager: BlobStorageManager | undefined;

function isBlobStorageEnabled(storageName: BlobStorageName) {
  const storageEnabledKey =
    `${storageName}_STORAGE_ENABLED` as keyof Environment;
  const storageEnabled = env[storageEnabledKey];

  return storageEnabled === true || storageEnabled === "true";
}

async function createBlobStorageManager() {
  const blobStorages = await Promise.all(
    Object.values(BLOB_STORAGE_NAMES).map(async (storageName) => {
      if (isBlobStorageEnabled(storageName)) {
        const [storage, storageError] = await createStorageFromEnv(storageName);

        if (storageError) {
          logger.warn(
            `${storageError.message}. Caused by: ${storageError.cause}`
          );
        }

        return storage;
      }
    })
  );

  const availableStorages = blobStorages.filter(
    (storage): storage is BlobStorage => !!storage
  );

  return new BlobStorageManager(availableStorages);
}

export async function getBlobStorageManager(): Promise<BlobStorageManager> {
  if (!blobStorageManager) {
    blobStorageManager = await createBlobStorageManager();
  }

  return blobStorageManager;
}
