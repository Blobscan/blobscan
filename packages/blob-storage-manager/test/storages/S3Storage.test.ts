import type { S3Client } from "@aws-sdk/client-s3";
import { beforeAll, describe, expect, it } from "vitest";

import { env } from "@blobscan/env";
import { testValidError } from "@blobscan/test";

import { S3Storage } from "../../src";
import { BlobStorageError } from "../../src/errors";
import type { S3StorageConfig } from "../../src/storages";
import { hexToBytes } from "../../src/utils";
import { NEW_BLOB_DATA, NEW_BLOB_HASH } from "../fixtures";

class S3StorageMock extends S3Storage {
  constructor(config: S3StorageConfig) {
    super(config);
  }

  closeMock() {
    // This would need to be implemented if we want to clean up test data
    return Promise.resolve();
  }

  healthCheck() {
    return super.healthCheck();
  }

  get bucketName(): string {
    return this._bucketName;
  }

  get s3Client(): S3Client {
    return this._s3Client;
  }
}

describe("S3Storage", () => {
  let storage: S3StorageMock;
  const expectedStoredBlobHash = "blobHash004";
  const expectedStoredBlobFileUri = `${
    env.CHAIN_ID
  }/${expectedStoredBlobHash.slice(2, 4)}/${expectedStoredBlobHash.slice(
    4,
    6
  )}/${expectedStoredBlobHash.slice(6, 8)}/${expectedStoredBlobHash.slice(
    2
  )}.bin`;
  const expectedStoredBlobData = "0x4fe40fc67f9c3a3ffa2be77d10fe7818";

  beforeAll(() => {
    if (!env.S3_STORAGE_BUCKET_NAME) {
      throw new Error("S3_STORAGE_BUCKET_NAME is not set");
    }

    storage = new S3StorageMock({
      chainId: env.CHAIN_ID,
      region: env.S3_STORAGE_REGION ?? "us-east-1",
      bucketName: env.S3_STORAGE_BUCKET_NAME,
      accessKeyId: env.S3_STORAGE_ACCESS_KEY_ID,
      secretAccessKey: env.S3_STORAGE_SECRET_ACCESS_KEY,
      endpoint: env.S3_STORAGE_ENDPOINT,
    });
  });

  it("should create a storage", async () => {
    expect(storage, "Storage should exist").toBeDefined();
    expect(storage.chainId, "Chain ID mismatch").toEqual(env.CHAIN_ID);
    expect(storage.bucketName, "Bucket name mismatch").toEqual(
      env.S3_STORAGE_BUCKET_NAME
    );
  });

  it("should return the correct uri given a blob hash", () => {
    const blobUri = storage.getBlobUri(expectedStoredBlobHash);

    expect(blobUri).toBe(expectedStoredBlobFileUri);
  });

  it("should return 'OK' if storage is healthy", async () => {
    await expect(storage.healthCheck()).resolves.toBe("OK");
  });

  it("should get a blob given its reference", async () => {
    // First store the blob so we can retrieve it
    await storage.storeBlob(expectedStoredBlobHash, expectedStoredBlobData);
    
    const blob = await storage.getBlob(expectedStoredBlobFileUri);

    expect(blob).toEqual(expectedStoredBlobData);
  });

  it("should remove a blob given its reference", async () => {
    // First store the blob so we can remove it
    await storage.storeBlob(expectedStoredBlobHash, expectedStoredBlobData);
    
    await storage.removeBlob(expectedStoredBlobFileUri);

    await expect(
      storage.getBlob(expectedStoredBlobFileUri)
    ).rejects.toThrowError();
  });

  it("should not throw an error when trying to remove a non-existent blob", async () => {
    await expect(
      storage.removeBlob("non-existent-blob-uri")
    ).resolves.not.toThrow();
  });

  it("should store a blob", async () => {
    const file = await storage.storeBlob(NEW_BLOB_HASH, NEW_BLOB_DATA);

    const blob = await storage.getBlob(file);

    expect(blob).toEqual(NEW_BLOB_DATA);
  });

  it("should return an uri when storing a blob", async () => {
    const file = await storage.storeBlob(NEW_BLOB_HASH, NEW_BLOB_DATA);

    expect(file).toEqual(
      `${env.CHAIN_ID}/01/00/ea/0100eac880c712dba4346c88ab564fa1b79024106f78f732cca49d8a68e4c174.bin`
    );
  });

  testValidError(
    "should throw a valid error if the bucket does not exist",
    async () => {
      const nonExistentBucket = "non-existent-bucket";

      await S3Storage.create({
        chainId: env.CHAIN_ID,
        region: env.S3_STORAGE_REGION ?? "us-east-1",
        bucketName: nonExistentBucket,
        endpoint: env.S3_STORAGE_ENDPOINT,
      });
    },
    Error,
    {
      checkCause: true,
    }
  );

  testValidError(
    "should throw a valid error when trying to get a non-existent blob",
    async () => {
      await storage.getBlob("missing-blob");
    },
    BlobStorageError,
    { checkCause: true }
  );

  testValidError(
    "should throw valid error if the bucket does not exist",
    async () => {
      const nonExistentBucket = "non-existent-bucket";

      const storage = await S3Storage.create({
        chainId: env.CHAIN_ID,
        region: env.S3_STORAGE_REGION ?? "us-east-1",
        bucketName: nonExistentBucket,
        endpoint: env.S3_STORAGE_ENDPOINT,
      });

      await storage.storeBlob(expectedStoredBlobHash, expectedStoredBlobData);
    },
    BlobStorageError,
    {
      checkCause: true,
    }
  );
});
