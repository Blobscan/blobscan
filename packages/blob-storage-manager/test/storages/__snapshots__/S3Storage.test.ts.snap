// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`S3Storage > should throw a valid error if the bucket does not exist 1`] = `"S3Storage failed: Creation failed: Storage is not reachable"`;

exports[`S3Storage > should throw a valid error if the bucket does not exist 2`] = `[Error: Bucket non-existent-bucket does not exist or is not accessible]`;

exports[`S3Storage > should throw a valid error when trying to get a non-existent blob 1`] = `"S3StorageMock failed: Failed to get blob with uri \\"missing-blob\\""`;

exports[`S3Storage > should throw a valid error when trying to get a non-existent blob 2`] = `[Error: no such key exists]`;

exports[`S3Storage > should throw valid error if the bucket does not exist 1`] = `"S3Storage failed: Creation failed: Storage is not reachable"`;

exports[`S3Storage > should throw valid error if the bucket does not exist 2`] = `[Error: Bucket non-existent-bucket does not exist or is not accessible]`;
